[dynamo] refactor frame skips and error messages in dynamo

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171358
* __->__ #170587

Fixes https://github.com/pytorch/pytorch/issues/170242.

There are a number of changes here to improve frame skip error messages:
- SkipFrame is no longer used by Dynamo to signify graph breaks that should skip the frame. SkipFrame should now only be used to unconditionally skip the frame without a graph break. The `skip_frame` attribute of `Unsupported` is now used for graph breaks to signal a frame skip.
- Uncaught `Unsupported`s (unhandled graph breaks) are now logged in symbolic_convert rather than convert_frame in order to improve logging consistency. (I kept the convert_frame call though since there are some `unimplemented` calls that occur outside of symbolic_convert).
- Chained graph breaks (graph break in loop, graph break in generic ctx manager, reconstruction errors) have better error messages

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

williamwen42 committed:
[dynamo] refactor frame skips in dynamo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor frame skips in dynamo"


Fixes https://github.com/pytorch/pytorch/issues/170242.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor frame skips in dynamo"


Fixes https://github.com/pytorch/pytorch/issues/170242.

There are a number of changes here to improve frame skip error messages:
- SkipFrame is no longer used by Dynamo to signify graph breaks that should skip the frame. SkipFrame should now only be used to unconditionally skip the frame without a graph break. The `skip_frame` attribute of `Unsupported` is now used for graph breaks to signal a frame skip.
- Uncaught `Unsupported`s (unhandled graph breaks) are now logged in symbolic_convert rather than convert_frame in order to improve logging consistency.
- Chained graph breaks (graph break in loop, graph break in generic ctx manager, reconstruction errors) have better error messages

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor frame skips and error messages in dynamo"


Fixes https://github.com/pytorch/pytorch/issues/170242.

There are a number of changes here to improve frame skip error messages:
- SkipFrame is no longer used by Dynamo to signify graph breaks that should skip the frame. SkipFrame should now only be used to unconditionally skip the frame without a graph break. The `skip_frame` attribute of `Unsupported` is now used for graph breaks to signal a frame skip.
- Uncaught `Unsupported`s (unhandled graph breaks) are now logged in symbolic_convert rather than convert_frame in order to improve logging consistency.
- Chained graph breaks (graph break in loop, graph break in generic ctx manager, reconstruction errors) have better error messages

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] refactor frame skips and error messages in dynamo"


Fixes https://github.com/pytorch/pytorch/issues/170242.

There are a number of changes here to improve frame skip error messages:
- SkipFrame is no longer used by Dynamo to signify graph breaks that should skip the frame. SkipFrame should now only be used to unconditionally skip the frame without a graph break. The `skip_frame` attribute of `Unsupported` is now used for graph breaks to signal a frame skip.
- Uncaught `Unsupported`s (unhandled graph breaks) are now logged in symbolic_convert rather than convert_frame in order to improve logging consistency.
- Chained graph breaks (graph break in loop, graph break in generic ctx manager, reconstruction errors) have better error messages

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

Lucaskabela commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1988,40 +1988,6 @@ def __call__(
-            # when we do not support graph breaks on bytecodes like LOAD_ATTR,
-            # BUILD_SET etc. In such case, we can fallback to eager without
-            # scaring users.
-            if soft_fail and graph_break_log.isEnabledFor(logging.DEBUG):
```
Can we add a TODO comment here and maybe an issue for easier tracking?

Lucaskabela reviewed (changes requested):
Some thoughts/questions here, but this is a very nice change :)

williamwen42 commented on test/dynamo/test_generator.py:
```diff
@@ -717,21 +717,18 @@ def whoo(t):
                 i += 1
                 yield t + j
 
-        @torch.compile(backend="eager", fullgraph=True)
```
This is because before, the graph break got wrapped up into a SkipFrame, which didn't actually bubble up as an error. Because I replaced SkipFrame with unimplemented(...) in a bunch of places, the graph break now properly bubbles up.

williamwen42 committed:
Update on "[dynamo] refactor frame skips and error messages in dynamo"


Fixes https://github.com/pytorch/pytorch/issues/170242.

There are a number of changes here to improve frame skip error messages:
- SkipFrame is no longer used by Dynamo to signify graph breaks that should skip the frame. SkipFrame should now only be used to unconditionally skip the frame without a graph break. The `skip_frame` attribute of `Unsupported` is now used for graph breaks to signal a frame skip.
- Uncaught `Unsupported`s (unhandled graph breaks) are now logged in symbolic_convert rather than convert_frame in order to improve logging consistency. (I kept the convert_frame call though since there are some `unimplemented` calls that occur outside of symbolic_convert).
- Chained graph breaks (graph break in loop, graph break in generic ctx manager, reconstruction errors) have better error messages

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

Lucaskabela reviewed (approved):
Thanks for addressing feedback :)

williamwen42 commented:
@pytorchbot merge
